File: tools/parity/crash-ladder.cpp
// The C++ half of the crash parity check (tests/crash.mjs).
//
// Prints the same ladder as tools/parity/crash-ladder.mjs, formatted
// identically, so the two builds can be diffed line for line. Compiled
// by a bare g++ — GRNSim.h needs no Unreal, which is the whole reason
// the solver lives there rather than in the pawn.
#include "GRNSim.h"
#include <cstdio>

// Signed zero is a formatting difference, not a numeric one: a parallel
// graze imparts exactly no rotation, and -0.0 and +0.0 are the same
// number. Adding 0.0 normalises the sign bit so the diff is about the
// model rather than about printf.
static double Z(double V) { return V + 0.0; }

int main()
{
	const double Full = GRNExact::CrashLatFull;
	const double Cases[][3] = {
		{ Full, 0.0, 1 }, { Full, 0.6, 1 }, { Full, -0.6, 1 },
		{ Full / 2, 0.6, 1 }, { Full * 4, -0.6, -1 }, { 2, 0.6, 1 },
	};
	for (const auto& C : Cases)
	{
		GRNSim::FImpact H = GRNSim::SolveWallImpact(C[0], C[1], C[2], 0.0);
		printf("wall into=%.3f hd=%+.2f side=%+.0f sev=%.6f yaw=%+.6f kick=%+.6f spin=%d nose=%d mul=%.6f slip=%+.6f hdg=%+.6f\n",
			C[0], C[1], C[2], Z(H.Severity), Z(H.Yaw), Z(H.Kick), (int)H.Spin, (int)H.NoseFirst,
			Z(H.SpeedMul), Z(H.SlipVel), Z(H.Heading));
	}
	for (int Fb = 0; Fb < 2; ++Fb)
	{
		GRNSim::FImpact T = GRNSim::SolveTrafficImpact(GRNExact::TrafficClosingFull, 0.2, 1, Fb != 0, 0.0);
		printf("traffic fromBehind=%d sev=%.6f yaw=%+.6f kick=%+.6f spin=%d\n", Fb, Z(T.Severity), Z(T.Yaw), Z(T.Kick), (int)T.Spin);
	}
	printf("scrape %.6f %.6f\n", GRNSim::ScrapeDrag(0.0), GRNSim::ScrapeDrag(1.0));
	return 0;
}

File: tools/parity/crash-ladder.mjs
// The ladder both builds are compared on. Kept beside the parity tools
// so the imports resolve the same way the rest of the suite's do.
import { HANDLING as H } from "../../src/game/handling.ts";
import { solveWallImpact, solveTrafficImpact, scrapeDrag } from "../../src/game/crash.ts";
const full = H.crashLatFull;
export const CASES = [[full,0,1],[full,0.6,1],[full,-0.6,1],[full/2,0.6,1],[full*4,-0.6,-1],[2,0.6,1]];
const f=(n,d)=>n.toFixed(d);
const sg=(n,d)=>`${n>=0?"+":""}${f(n,d)}`;
export function ladder() {
  const out = [];
  for (const [into,hd,side] of CASES) {
    const h = solveWallImpact({into,heading:hd,side,crashResist:0});
    out.push(`wall into=${f(into,3)} hd=${sg(hd,2)} side=${sg(side,0)} sev=${f(h.severity,6)} yaw=${sg(h.yaw,6)} kick=${sg(h.kick,6)} spin=${h.spin?1:0} nose=${h.noseFirst?1:0} mul=${f(h.speedMul,6)} slip=${sg(h.slipVel,6)} hdg=${sg(h.heading,6)}`);
  }
  for (const fb of [false,true]) {
    const t = solveTrafficImpact({closing:H.trafficClosingFull,heading:0.2,shove:1,fromBehind:fb,crashResist:0});
    out.push(`traffic fromBehind=${fb?1:0} sev=${f(t.severity,6)} yaw=${sg(t.yaw,6)} kick=${sg(t.kick,6)} spin=${t.spin?1:0}`);
  }
  out.push(`scrape ${f(scrapeDrag(0),6)} ${f(scrapeDrag(1),6)}`);
  return out;
}
if (import.meta.url === `file://${process.argv[1]}`) console.log(ladder().join("\n"));
